File: docs/onepay/test-script.go
/*
OnePay PayOut — Account check + Funds transfer test script.

Based on the official OnePay sample code and your Python test script.

Run:
    go run main.go
*/

package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

// ---------------------------------------------------------------------------
// Credentials (MTF sandbox)
// ---------------------------------------------------------------------------
const (
	PartnerID  = "TESTVFICPO"
	PartnerKey = "C2B5DA903DAE19E215454211C66A59DD"
	AccountID  = "666894931888"
	BaseURL    = "https://mtf.onepay.vn"

	// Happy-case test recipient from the OnePay docs
	TestSwiftCode     = "ICBVVNVX"      // VietinBank (from official sample)
	TestAccountNumber = "103000614434"  // from official sample
	TestHolderName    = "Nguyen Danh Hoang"
	TestAmount         = "10000"         // min 10,000 VND
)

// ---------------------------------------------------------------------------
// Signature constants — matching official Authorization specifications
// ---------------------------------------------------------------------------
const (
	Scheme      = "OWS1"
	Algorithm   = "OWS1-HMAC-SHA256"
	Terminator  = "ows1_request"
	OwsRegion   = "onepay"
	OwsService  = "onepayout"
	DateLayout  = "20060102T150405Z"
	EmptyBodySha = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
)

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func hmacSHA256(key []byte, msg string) []byte {
	h := hmac.New(sha256.New, key)
	h.Write([]byte(msg))
	return h.Sum(nil)
}

func sha256Hex(data []byte) string {
	h := sha256.New()
	h.Write(data)
	return hex.EncodeToString(h.Sum(nil))
}

// uriEncode mimics Python's urllib.parse.quote(..., safe="~" or "~/")
func uriEncode(data string, encodeSlash bool) string {
	var result strings.Builder
	for i := 0; i < len(data); i++ {
		c := data[i]
		// Safe characters: alphanumeric, '-', '_', '.', '~'
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') ||
			c == '-' || c == '_' || c == '.' || c == '~' {
			result.WriteByte(c)
		} else if c == '/' {
			if encodeSlash {
				result.WriteString("%2F")
			} else {
				result.WriteByte('/')
			}
		} else {
			result.WriteString(fmt.Sprintf("%%%02X", c))
		}
	}
	return result.String()
}

// ---------------------------------------------------------------------------
// Signing Core Logic
// ---------------------------------------------------------------------------

func signRequest(
	accessKeyID string,
	secretKey string,
	region string,
	service string,
	httpMethod string,
	uri string,
	queryParameters map[string]string,
	signedHeaders map[string]string,
	payload []byte,
	timestamp time.Time,
) string {
	// 1. Canonical URI
	canonicalURI := uriEncode(uri, false)

	// 2. Canonical Query String — sort by raw key, then encode key=value
	keys := make([]string, 0, len(queryParameters))
	for k := range queryParameters {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var canonicalQueryParts []string
	for _, k := range keys {
		part := uriEncode(k, true) + "=" + uriEncode(queryParameters[k], true)
		canonicalQueryParts = append(canonicalQueryParts, part)
	}
	canonicalQueryString := strings.Join(canonicalQueryParts, "&")

	// 3. Canonical Headers + SignedHeaders
	headerKeys := make([]string, 0, len(signedHeaders))
	for k := range signedHeaders {
		headerKeys = append(headerKeys, k)
	}
	// Sort header names in lowercase
	sort.Slice(headerKeys, func(i, j int) bool {
		return strings.ToLower(headerKeys[i]) < strings.ToLower(headerKeys[j])
	})

	var canonicalHeaders strings.Builder
	var signedHeaderNames []string
	for _, k := range headerKeys {
		lowerKey := strings.ToLower(k)
		val := strings.TrimSpace(signedHeaders[k])
		canonicalHeaders.WriteString(lowerKey + ":" + val + "\n")
		signedHeaderNames = append(signedHeaderNames, lowerKey)
	}
	joinedSignedHeaderNames := strings.Join(signedHeaderNames, ";")

	// 4. Hashed Payload
	var hashedPayload string
	if payload != nil {
		if len(payload) > 0 {
			hashedPayload = sha256Hex(payload)
		} else {
			hashedPayload = EmptyBodySha
		}
	} else {
		hashedPayload = "UNSIGNED-PAYLOAD"
	}

	// 5. Canonical Request
	canonicalRequest := strings.Join([]string{
		httpMethod,
		canonicalURI,
		canonicalQueryString,
		canonicalHeaders.String(),
		joinedSignedHeaderNames,
		hashedPayload,
	}, "\n")

	// 6. StringToSign
	tsISO := timestamp.UTC().Format(DateLayout)
	tsDate := timestamp.UTC().Format("20060102")
	scope := tsDate + "/" + region + "/" + service + "/" + Terminator
	stringToSign := strings.Join([]string{
		Algorithm,
		tsISO,
		scope,
		sha256Hex([]byte(canonicalRequest)),
	}, "\n")

	// 7. Signing Key — 4 layers: date → region → service → terminator
	dateKey := hmacSHA256([]byte(Scheme+secretKey), tsDate)
	dateRegionKey := hmacSHA256(dateKey, region)
	dateRegionServiceKey := hmacSHA256(dateRegionKey, service)
	signingKey := hmacSHA256(dateRegionServiceKey, Terminator)

	// 8. Signature
	signature := hex.EncodeToString(hmacSHA256(signingKey, stringToSign))

	credential := accessKeyID + "/" + scope

	fmt.Println("===== CanonicalRequest =====")
	fmt.Printf("%q\n", canonicalRequest)
	fmt.Println("===== StringToSign =====")
	fmt.Printf("%q\n", stringToSign)
	fmt.Println("===== Signing Key =====")
	fmt.Println(hex.EncodeToString(signingKey))
	fmt.Println("===== Signature =====")
	fmt.Println(signature)

	return fmt.Sprintf("%s Credential=%s,SignedHeaders=%s,Signature=%s",
		Algorithm, credential, joinedSignedHeaderNames, signature)
}

// sendRequest prints the RAW request metadata and body, executes it, and handles response
func sendRequest(method, requestURL string, headers map[string]string, body []byte) (*http.Response, []byte, error) {
	var bodyReader io.Reader
	if body != nil {
		bodyReader = bytes.NewReader(body)
	}

	req, err := http.NewRequest(method, requestURL, bodyReader)
	if err != nil {
		return nil, nil, err
	}

	// Set headers
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	fmt.Println()
	fmt.Println("========== RAW HTTP REQUEST ==========")
	fmt.Printf("%s %s %s\n", req.Method, req.URL.RequestURI(), req.Proto)
	for k, vs := range req.Header {
		for _, v := range vs {
			fmt.Printf("%s: %s\n", k, v)
		}
	}
	if len(body) > 0 {
		fmt.Println()
		fmt.Println(string(body))
	}
	fmt.Println("=======================================")
	fmt.Println()

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, nil, err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, nil, err
	}

	fmt.Printf("Status: %d\n", resp.StatusCode)

	// Format response nicely if JSON
	var prettyJSON bytes.Buffer
	if err := json.Indent(&prettyJSON, respBody, "", "  "); err == nil {
		fmt.Println(prettyJSON.String())
	} else {
		fmt.Println(string(respBody))
	}

	return resp, respBody, nil
}

// ---------------------------------------------------------------------------
// 1. Account check (GET /customers)
// ---------------------------------------------------------------------------
func testCheckAccount() (*http.Response, []byte, error) {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("TEST 1: Account Check (GET /customers)")
	fmt.Println(strings.Repeat("=", 60))

	ts := time.Now().UTC()
	tsISO := ts.Format(DateLayout)
	requestID := fmt.Sprintf("REQ%d", time.Now().Unix())

	headerSign := map[string]string{
		"Accept":       "application/json",
		"Content-Type": "application/json",
		"X-OP-Date":    tsISO,
		"X-OP-Expires": "3600",
	}

	queryParamMap := map[string]string{
		"request_id":     requestID,
		"swift_code":     TestSwiftCode,
		"account_number": TestAccountNumber,
		"amount":         TestAmount,
		"account_id":     AccountID,
	}

	// Sign signatures using custom canonical-order map values
	queryAfter := "?"
	keys := make([]string, 0, len(queryParamMap))
	for k := range queryParamMap {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	for _, k := range keys {
		queryAfter += k + "=" + url.QueryEscape(queryParamMap[k]) + "&"
	}
	queryAfter = strings.TrimSuffix(queryAfter, "&")

	uri := "/onepayout/api/v1/customers"

	authString := signRequest(
		PartnerID,
		PartnerKey,
		OwsRegion,
		OwsService,
		"GET",
		uri,
		queryParamMap,
		headerSign,
		[]byte(""),
		ts,
	)

	urlRequest := BaseURL + uri + queryAfter
	headerRequest := map[string]string{
		"Accept":             "application/json",
		"Content-Type":       "application/json",
		"X-OP-Date":          tsISO,
		"X-OP-Authorization": authString,
		"X-OP-Expires":       "3600",
	}

	return sendRequest("GET", urlRequest, headerRequest, nil)
}

// ---------------------------------------------------------------------------
// 2. Funds Transfer (PUT /accounts/{id}/funds_transfers/{id})
// ---------------------------------------------------------------------------
func testFundsTransfer() (*http.Response, []byte, error) {
	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("TEST 2: Funds Transfer (PUT /accounts/.../funds_transfers/...)")
	fmt.Println(strings.Repeat("=", 60))

	ts := time.Now().UTC()
	tsISO := ts.Format(DateLayout)
	fundsTransferID := fmt.Sprintf("TF%d", time.Now().Unix())

	uri := fmt.Sprintf("/onepayout/api/v1/accounts/%s/funds_transfers/%s", AccountID, fundsTransferID)

	type TransferPayload struct {
		SwiftCode         string `json:"swift_code"`
		AccountNumber     string `json:"account_number"`
		HolderName        string `json:"holder_name"`
		Amount            string `json:"amount"`
		Currency          string `json:"currency"`
		FundsTransferInfo string `json:"funds_transfer_info"`
		Remark            string `json:"remark"`
	}

	payload := TransferPayload{
		SwiftCode:         TestSwiftCode,
		AccountNumber:     TestAccountNumber,
		HolderName:        TestHolderName,
		Amount:            TestAmount,
		Currency:          "VND",
		FundsTransferInfo: fundsTransferID,
		Remark:            "test transfer from script",
	}

	bodyBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, nil, err
	}

	headerSign := map[string]string{
		"Accept":       "application/json",
		"Content-Type": "application/json",
		"X-OP-Date":    tsISO,
		"X-OP-Expires": "3600",
	}

	authString := signRequest(
		PartnerID,
		PartnerKey,
		OwsRegion,
		OwsService,
		"PUT",
		uri,
		nil, // empty query parameters
		headerSign,
		bodyBytes,
		ts,
	)

	urlRequest := BaseURL + uri
	headerRequest := map[string]string{
		"Accept":             "application/json",
		"Content-Type":       "application/json",
		"X-OP-Date":          tsISO,
		"X-OP-Authorization": authString,
		"X-OP-Expires":       "3600",
	}

	return sendRequest("PUT", urlRequest, headerRequest, bodyBytes)
}

// ---------------------------------------------------------------------------
// Main Flow
// ---------------------------------------------------------------------------
func main() {
	resp, respBody, err := testCheckAccount()
	if err != nil {
		fmt.Printf("\n❌ Account check failed with application error: %v\n", err)
		return
	}

	if resp.StatusCode == 200 {
		var data map[string]interface{}
		if err := json.Unmarshal(respBody, &data); err != nil {
			fmt.Printf("\n❌ Failed to parse JSON response: %v\n", err)
			return
		}

		responseCode, _ := data["response_code"].(string)
		message, _ := data["message"].(string)

		if responseCode == "00" {
			fmt.Println("\n✅ Account check passed, proceeding to transfer...")
			_, _, err := testFundsTransfer()
			if err != nil {
				fmt.Printf("\n❌ Funds transfer failed with application error: %v\n", err)
			}
		} else {
			fmt.Printf("\n❌ Account check failed: %s %s\n", responseCode, message)
		}
	} else {
		fmt.Printf("\n❌ Account check HTTP error: %d\n", resp.StatusCode)
	}
}

File: docs/onepay/test-script.py
"""
OnePay PayOut — Account check + Funds transfer test script.

Based on the official OnePay sample code at:
  docs/onepay/OnePayOWS/Python/

Run:
    python3 test-script.py
"""

import hashlib
import hmac
import time
import json
import urllib.parse

import requests

# ---------------------------------------------------------------------------
# Credentials (MTF sandbox)
# ---------------------------------------------------------------------------
PARTNER_ID  = "TESTVFICPO"
PARTNER_KEY = "C2B5DA903DAE19E215454211C66A59DD"
ACCOUNT_ID  = "666894931888"
BASE_URL    = "https://mtf.onepay.vn"

# Happy-case test recipient from the OnePay docs
TEST_SWIFT_CODE     = "ICBVVNVX"      # VietinBank (from official sample)
TEST_ACCOUNT_NUMBER = "103000614434"  # from official sample
TEST_HOLDER_NAME    = "Nguyen Danh Hoang"
TEST_AMOUNT         = "10000"         # min 10,000 VND

# ---------------------------------------------------------------------------
# Signature constants — matching official Authorization.py
# ---------------------------------------------------------------------------
SCHEME      = "OWS1"
ALGORITHM   = "OWS1-HMAC-SHA256"
TERMINATOR  = "ows1_request"
OWS_REGION  = "onepay"
OWS_SERVICE = "onepayout"
DATE_LAYOUT = "%Y%m%dT%H%M%SZ"

EMPTY_BODY_SHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"


# ---------------------------------------------------------------------------
# Helpers — matching official Util.py + Authorization.py
# ---------------------------------------------------------------------------
def hmac_sha256(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def uri_encode(data: str, encode_slash: bool) -> str:
    if encode_slash:
        return urllib.parse.quote(data.encode("utf-8"), safe="~")
    else:
        return urllib.parse.quote(data.encode("utf-8"), safe="~/")


def get_bytes(value: str) -> bytes:
    return value.encode("utf-8")


# ---------------------------------------------------------------------------
# Signing — exactly matching official Authorization.py sign()
# ---------------------------------------------------------------------------
def sign_request(
    access_key_id: str,
    secret_key: str,
    region: str,
    service: str,
    http_method: str,
    uri: str,
    query_parameters: dict,
    signed_headers: dict,
    payload: bytes,
    timestamp: time.struct_time,
) -> str:
    # 1. Canonical URI
    canonical_uri = uri_encode(uri, False)

    # 2. Canonical Query String — sort by raw key, then encode key=value
    canonical_query_string = ""
    for key, value in sorted(query_parameters.items(), key=lambda kv: kv[0]):
        if len(canonical_query_string) > 0:
            canonical_query_string += "&"
        canonical_query_string += uri_encode(str(key), True) + "=" + uri_encode(str(value), True)

    # 3. Canonical Headers + SignedHeaders
    canonical_headers = ""
    signed_header_names = ""
    for key in sorted(signed_headers.keys(), key=lambda k: k.lower()):
        canonical_headers += key.lower() + ":" + str(signed_headers[key]).strip() + "\n"
        if len(signed_header_names) > 0:
            signed_header_names += ";"
        signed_header_names += key.lower()

    # 4. Hashed Payload — matching official logic
    if payload is not None:
        if len(payload) > 0:
            hashed_payload = sha256_hex(payload)
        else:
            hashed_payload = EMPTY_BODY_SHA256
    else:
        hashed_payload = "UNSIGNED-PAYLOAD"

    # 5. Canonical Request
    canonical_request = (
        http_method + "\n"
        + canonical_uri + "\n"
        + canonical_query_string + "\n"
        + canonical_headers + "\n"
        + signed_header_names + "\n"
        + hashed_payload
    )

    # 6. StringToSign
    ts_iso = time.strftime(DATE_LAYOUT, timestamp)
    ts_date = time.strftime("%Y%m%d", timestamp)
    scope = ts_date + "/" + region + "/" + service + "/" + TERMINATOR
    string_to_sign = (
        ALGORITHM + "\n"
        + ts_iso + "\n"
        + scope + "\n"
        + sha256_hex(get_bytes(canonical_request))
    )

    # 7. Signing Key — 4 layers: date → region → service → terminator
    date_key = hmac_sha256(get_bytes(SCHEME + secret_key), ts_date)
    date_region_key = hmac_sha256(date_key, region)
    date_region_service_key = hmac_sha256(date_region_key, service)
    signing_key = hmac_sha256(date_region_service_key, TERMINATOR)

    # 8. Signature
    signature = hmac_sha256(signing_key, string_to_sign).hex()

    credential = access_key_id + "/" + scope

    print("===== CanonicalRequest =====")
    print(repr(canonical_request))
    print("===== StringToSign =====")
    print(repr(string_to_sign))
    print("===== Signing Key =====")
    print(signing_key.hex())
    print("===== Signature =====")
    print(signature)

    return (
        ALGORITHM
        + " Credential=" + credential
        + ",SignedHeaders=" + signed_header_names
        + ",Signature=" + signature
    )


def _send_request(method, url, headers, body=None):
    """Send a request and print the raw HTTP exchange + response."""
    req = requests.Request(method, url, headers=headers, data=body)
    prepared = req.prepare()

    print()
    print("========== RAW HTTP REQUEST ==========")
    print(f"{prepared.method} {prepared.url} HTTP/1.1")
    for k, v in prepared.headers.items():
        print(f"{k}: {v}")
    if body:
        print()
        print(body.decode("utf-8") if isinstance(body, bytes) else body)
    print("=======================================")
    print()

    resp = requests.Session().send(prepared, timeout=15)
    print(f"Status: {resp.status_code}")
    try:
        print(json.dumps(resp.json(), indent=2, ensure_ascii=False))
    except ValueError:
        print(resp.text)
    return resp


# ---------------------------------------------------------------------------
# 1. GET /onepayout/api/v1/customers — Account check
# ---------------------------------------------------------------------------
def test_check_account():
    print("\n" + "=" * 60)
    print("TEST 1: Account Check (GET /customers)")
    print("=" * 60)

    ts = time.gmtime()
    ts_iso = time.strftime(DATE_LAYOUT, ts)
    request_id = f"REQ{int(time.time())}"

    header_sign = {
        "Accept":       "application/json",
        "Content-Type": "application/json",
        "X-OP-Date":    ts_iso,
        "X-OP-Expires": "3600",
    }

    query_param_map = {
        "request_id":     request_id,
        "swift_code":     TEST_SWIFT_CODE,
        "account_number": TEST_ACCOUNT_NUMBER,
        "amount":         TEST_AMOUNT,
        "account_id":     ACCOUNT_ID,
    }

    query_params_encoded = {}
    query_after = "?"
    for key, value in sorted(query_param_map.items(), key=lambda kv: kv[0]):
        query_params_encoded[key] = urllib.parse.quote(str(value), encoding="utf-8")
        query_after += key + "=" + urllib.parse.quote(str(value), encoding="utf-8") + "&"
    query_after = query_after.rstrip("&")

    uri = "/onepayout/api/v1/customers"

    auth_string = sign_request(
        access_key_id=PARTNER_ID,
        secret_key=PARTNER_KEY,
        region=OWS_REGION,
        service=OWS_SERVICE,
        http_method="GET",
        uri=uri,
        query_parameters=query_params_encoded,
        signed_headers=header_sign,
        payload=get_bytes(""),
        timestamp=ts,
    )

    url_request = BASE_URL + uri + query_after
    header_request = {
        "Accept":             "application/json",
        "Content-Type":       "application/json",
        "X-OP-Date":          ts_iso,
        "X-OP-Authorization": auth_string,
        "X-OP-Expires":       "3600",
    }

    return _send_request("GET", url_request, header_request)


# ---------------------------------------------------------------------------
# 2. PUT /onepayout/api/v1/accounts/{id}/funds_transfers/{id} — Transfer
# ---------------------------------------------------------------------------
def test_funds_transfer():
    print("\n" + "=" * 60)
    print("TEST 2: Funds Transfer (PUT /accounts/.../funds_transfers/...)")
    print("=" * 60)

    ts = time.gmtime()
    ts_iso = time.strftime(DATE_LAYOUT, ts)
    funds_transfer_id = f"TF{int(time.time())}"

    uri = f"/onepayout/api/v1/accounts/{ACCOUNT_ID}/funds_transfers/{funds_transfer_id}"

    body = json.dumps({
        "swift_code":         TEST_SWIFT_CODE,
        "account_number":     TEST_ACCOUNT_NUMBER,
        "holder_name":        TEST_HOLDER_NAME,
        "amount":             TEST_AMOUNT,
        "currency":           "VND",
        "funds_transfer_info": funds_transfer_id,
        "remark":             "test transfer from script",
    })
    body_bytes = get_bytes(body)

    header_sign = {
        "Accept":       "application/json",
        "Content-Type": "application/json",
        "X-OP-Date":    ts_iso,
        "X-OP-Expires": "3600",
    }

    auth_string = sign_request(
        access_key_id=PARTNER_ID,
        secret_key=PARTNER_KEY,
        region=OWS_REGION,
        service=OWS_SERVICE,
        http_method="PUT",
        uri=uri,
        query_parameters={},
        signed_headers=header_sign,
        payload=body_bytes,
        timestamp=ts,
    )

    url_request = BASE_URL + uri
    header_request = {
        "Accept":             "application/json",
        "Content-Type":       "application/json",
        "X-OP-Date":          ts_iso,
        "X-OP-Authorization": auth_string,
        "X-OP-Expires":       "3600",
    }

    return _send_request("PUT", url_request, header_request, body=body_bytes)


if __name__ == "__main__":
    resp = test_check_account()
    if resp.status_code == 200:
        data = resp.json()
        if data.get("response_code") == "00":
            print("\n✅ Account check passed, proceeding to transfer...")
            test_funds_transfer()
        else:
            print(f"\n❌ Account check failed: {data.get('response_code')} {data.get('message')}")
    else:
        print(f"\n❌ Account check HTTP error: {resp.status_code}")
